windows/devices/uwb/UwbSession.cxx

```diff
@@ -261,6 +239,58 @@ UwbSession::GetApplicationConfigurationParametersImpl()
     }
 }
 
+std::vector<UwbApplicationConfigurationParameter>
+UwbSession::GetApplicationConfigurationParametersImpl(std::vector<::uwb::protocol::fira::UwbApplicationConfigurationParameterType> requestedTypes)
+{
+    uint32_t sessionId = GetId();
+
+    auto resultFuture = m_uwbSessionConnector->GetApplicationConfigurationParameters(sessionId, requestedTypes);
+    if (!resultFuture.valid()) {
+        PLOG_ERROR << "failed to obtain application configuration parameters for session id " << sessionId;
+        throw UwbException(UwbStatusGeneric::Failed);
+    }
+
+    try {
+        auto [uwbStatus, applicationConfigurationParameters] = resultFuture.get();
+        if (!IsUwbStatusOk(uwbStatus)) {
+            throw UwbException(uwbStatus);
+        }
+        return applicationConfigurationParameters;
+    } catch (UwbException &uwbException) {
+        PLOG_ERROR << "caught exception attempting to obtain application configuration parameters for session id " << sessionId << " (" << ToString(uwbException.Status) << ")";
+        throw uwbException;
+    } catch (std::exception &e) {
+        PLOG_ERROR << "caught unexpected exception attempting to obtain application configuration parameters for session id " << sessionId << " (" << e.what() << ")";
+        throw e;
+    }
+}
+
+void
+UwbSession::SetApplicationConfigurationParametersImpl(std::vector<::uwb::protocol::fira::UwbApplicationConfigurationParameter> params)
+{
+    uint32_t sessionId = GetId();
+    while (true) {
```

Comment:
Why is this loop here? It looks like the idea is to keep trying to set the parameters if `IsUwbStatusRetry` is true, but when would that ever be the case? From my understanding, when we send the SetApplicationConfigurationParameters command to the device, it should return either ok, invalid parameter, invalid range, or error session active.

Reply:
After looking at the UCI spec again, I agree with what you're saying here. That being said, we don't currently handle CORE_GENERIC_ERROR_NTF from the UWBS, especially not when the Status is set to STATUS_COMMAND_RETRY. Opening #246 to track it

Reply:
Even though the spec explicitly calls out the status values for `SESSION_SET_APP_CONFIG_RSP` and retry is _not_ included, I think `STATUS_COMMAND_RETRY` can apply to any command. I'd rather handle it than not.

However, I'd also say that we should handle all such status indications the same way, and have configurable settings controlling it such as maximum number of retries, delay between retries, etc. I'd also rather see that done in a separate PR in the future when we're more focused on reliability. Let's just fail if the status is retry for now. Created #247 to track that work.

Reply:
I see, since `STATUS_COMMAND_RETRY` is technically a possible value, then it should be handled. I just don't know when that would ever be the case, since it is never mentioned anywhere in the UCI spec (besides the list of status codes).

I definitely agree with having a max number of retries, since an infinite loop could be very bad. Even if it's likely that we're asked to retry a command an infinite number of times, it's still good to code defensively

Reply:
I think we should leave this to a future PR. For now, I'll have the function return an error when it requests retry through the status, as it's consistent with how the other functions handle it

Reply:
> I see. Because `STATUS_COMMAND_RETRY` is technically a possible value, then it should be handled. I just don't know when that would ever be the case, since it is never mentioned anywhere in the UCI spec (besides the list of status codes).

Given the spec is often ambiguous, my guess is that they didn't explicitly call out all the places `STATUS_COMMAND_RETRY` could be returned, and that it could be returned for any command. From my experience, it's not uncommon for simple hardware to return a "retry" status indication when it just can't handle the request at that specific moment, instead of implementing additional complex logic in the hardware/firmware to handle queuing/control flow of concurrent operations when they can only support sequential (this happens a lot with microcontrollers for cost-savings).

> I definitely agree with having a max number of retries, since an infinite loop could be very bad. Even if it's likely that we're asked to retry a command an infinite number of times, it's still good to code defensively

💯